Add Plasma Mobile Gear 21.05

samueldr opened the pull request:
## Motivation for this change

#121345 is a mess to review because it is just plain too big. I'll rebase #121345 on top of this PR, which will make it hold **only** the Plasma Mobile *shell*. (The "phone GUI environment".)

Those applications are independent from *actual* plasma mobile.

They can seemingly be used (with caveats) on e.g. Phosh.

*Only applications with a release tags are packaged.*

Note that these applications launch just right on my X11 desktop with Awesome WM. No real caveats there AFAICT.

:heavy_check_mark: This PR is **ready** for reviewing.

 - https://www.plasma-mobile.org/2021/05/10/plasma-mobile-gear-may-2021/
 - https://www.plasma-mobile.org/info/plasma-mobile-gear-21-05/

* * *

### Caveats

None of those caveats are bugs in the packaging of these applications.

<details><summary>(Long scary list, not actual packaging bugs)</summary>

#### Icons wont work as expected

 - https://bugs.kde.org/show_bug.cgi?id=435236
 - Workaround: https://github.com/samueldr/hack-merged-icons

```nix
{
  # Work around issues with plasma mobile apps on phosh
  #  - https://bugs.kde.org/show_bug.cgi?id=435236
  hacks.mergedIcons = {
    enable = true;

    # Helps with breeze fallback icons under gnome/phosh
    # Otherwise it might use Adwaita
    removeLegacyIcons = true;
    disabledThemes = [
      "Adwaita"
    ];

    # Options that aren't used
    #masqueradeAs = [
    #  "Adwaita" # Doesn't help really
    #];
    #iconThemes = [
    #];
  };
}
```

#### Will not start by default

By default the apps on Phosh (or GNOME too I suppose?) won't start:

> ```
> Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland ro tun on Wayland anyway.
> ```

They'll try to start using `xcb`/X11 and crash. How rude.

The workaround is to force Wayland... Rude...

```nix

{
  environment.variables = {
    QT_QPA_PLATFORM = "wayland";
  };
}
```

It is unclear why those apps are willfully ignoring starting as wayland apps in Phosh (and GNOME I suppose).

#### CSDs in Phosh

While not really an issue, one can *completely disable Qt CSDs* this way:

```nix

{
  environment.variables = {
    QT_WAYLAND_DISABLE_WINDOWDECORATION = "1";
  };
}
```

Seems to me there is a need for some coordination between "phone environments" to properly tell when CSDs are needed and when they're not for docked/undocked modes :/.

</details>

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Actual tests:

I ran applications on my X11 desktop **and** under Phosh. Apps were *also* ran in their previous releases under Plasma Mobile, but not this most up-to-date branch yet.

**Fully tested:**

 - alligator
 - calindori
 - kalk
 - kclock
 - koko
 - krecorder
 - ktrip
 - ~~kweather~~ (not part of 21.05)
 - ~~plasma-camera~~ (not part of 21.05)
 - plasma-phonebook

**Features untested:**

 - plasma-dialer (dialing/receiving a call... basically ofono integration)
 - spacebar (sms receiving/sending... basically ofono integration)

cc @NixOS/qt-kde 

#### TODO

 - [x] Use an update script for Plasma Mobile Gear

samueldr committed:
kalk: init at 21.05

samueldr committed:
kclock: init at 21.05

samueldr committed:
alligator: init at 21.05

samueldr committed:
calindori: init at 21.05

samueldr committed:
krecorder: init at 21.05

samueldr committed:
plasma-dialer: init at 21.05

samueldr committed:
plasma-phonebook: init at 21.05

samueldr committed:
spacebar: init at 21.05

samueldr committed:
libqofono: init at 0.103

samueldr committed:
kirigami-addons: init at 21.05

r-rmcgibbo commented:
Result of `nixpkgs-review pr 125156` at 605cb65f run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages skipped due to time constraints:</summary>
  <ul>
    <li>kclock</li>
    <li>kweather</li>
  </ul>
</details>
<details>
  <summary>10 packages built successfully:</summary>
  <ul>
    <li>alligator</li>
    <li>calindori</li>
    <li>kalk</li>
    <li>kirigami-addons</li>
    <li>krecorder</li>
    <li>libqofono</li>
    <li>plasma-camera</li>
    <li>plasma-dialer</li>
    <li>plasma-phonebook</li>
    <li>spacebar-sms</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/tools/networking/libqofono/default.nix:28:5:
```
   |
28 |     ./0001-NixOS-Skip-tests-they-re-shock-full-of-hardcoded-FHS.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/tools/networking/libqofono/default.nix:24:5:
```
   |
24 |     (substituteAll {
   |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/applications/science/math/kalk/default.nix:4:3:
```
  |
4 | , fetchpatch
  |   ^
```
</li>
  </ul>
</details>

dotlambda commented on pkgs/applications/networking/feedreaders/alligator/default.nix:
```diff
@@ -0,0 +1,50 @@
+    homepage = "https://invent.kde.org/plasma-mobile/alligator";
+    # https://invent.kde.org/plasma-mobile/alligator/-/commit/db30f159c4700244532b17a260deb95551045b7a
+    #  * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
+    license = licenses.gpl3Plus;
```
This doesn't make sense.

samueldr replied:
You're right.

dotlambda commented on pkgs/applications/graphics/plasma-camera/default.nix:
```diff
@@ -0,0 +1,60 @@
+  nativeBuildInputs = [
+    cmake
+    extra-cmake-modules
+    wrapGAppsHook
```
This will double wrap. Better add to `qtWrapperArgs`.

samueldr replied:
I'm not convinced that it's better. In fact I'm not convinced peppering wrappers everywhere is a solution to any kind of problem we're having here :/.

I'm not sure what to add, but I assume we'd copy some of the logic of `wrapGAppsHook` in `qtWrapperArgs`. Wouldn't that risk drifting from the main implementation and maybe break?

Any documentation for combining `GAppsHook` and `qtWrapper`?

dotlambda replied:
`qtWrapperArgs+=("''${gappsWrapperArgs[@]}")` should do the trick.

SuperSandro2000 commented on pkgs/applications/misc/kweather/default.nix:
```diff
@@ -0,0 +1,50 @@
+    domain = "invent.kde.org";
+    owner = "plasma-mobile";
+    repo = pname;
+    rev = "${version}";
```
```suggestion
    rev = version;
```

SuperSandro2000 commented on pkgs/applications/misc/kweather/default.nix:
```diff
@@ -0,0 +1,50 @@
+  meta = with lib; {
+    description = "Weather application for Plasma Mobile";
+    homepage = "https://invent.kde.org/plasma-mobile/kweather";
+    license = [ licenses.gpl2Plus licenses.cc-by-40 ];
```
```suggestion
    license = with licenses; [ gpl2Plus cc-by-40 ];
```

samueldr commented:
I intend to merge this by the end of this week. (Unless someone else merges this beforehand.)

dotlambda commented on pkgs/applications/graphics/plasma-camera/default.nix:
```diff
@@ -0,0 +1,65 @@
+mkDerivation rec {
+  pname = "plasma-camera";
+  # The last tagged version is more than a year old and slightly broken
+  version = "2021-03-02";
```
```suggestion
  version = "unstable-2021-03-02";
```

dotlambda commented on pkgs/tools/networking/libqofono/default.nix:
```diff
@@ -0,0 +1,58 @@
+  patches = [
+    (substituteAll {
+      src = ./0001-NixOS-provide-mobile-broadband-provider-info-path.patch;
+      mobile_broadband_provider_info = mobile-broadband-provider-info;
```
I would prefer changing the name in the patches to use
```suggestion
      inherit mobile-broadband-provider-info;
```
but that's up to you.

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5997,6 +6001,8 @@ in
 
   kalibrate-hackrf = callPackage ../applications/radio/kalibrate-hackrf { };
 
+  kalk = libsForQt5.callPackage ../applications/science/math/kalk { };
```
should be below kakoune.

dotlambda replied:
should be below kakoune

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -24627,6 +24641,8 @@ in
 
   kile = libsForQt5.callPackage ../applications/editors/kile { };
 
+  kirigami-addons = libsForQt5.callPackage ../development/libraries/kirigami-addons { };
```
Shouldn't this be in `pkgs/top-level/qt5-packages.nix`?

dotlambda commented:
Running `kclock` prints
> QML debugging is enabled. Only use this in a safe environment.

Is there a way to make a release build without that?

samueldr commented:
https://invent.kde.org/plasma-mobile/kclock/-/commit/64931326ee991e9fc24f749944cdf6815c7426e3

*sighs*

samueldr commented:
I believe everything @dotlambda requested has been changed.

About inclusion in `pkgs/top-level/qt5-packages.nix`... Really it just wasn't clear to me where things should go for a lot of this.

dotlambda commented:
> About inclusion in `pkgs/top-level/qt5-packages.nix`... Really it just wasn't clear to me where things should go for a lot of this.

I would say: everything that might be used with multiple versions of Qt, i.e. libraries, go in `qt5-packages.nix`.

samueldr commented:
:x: **This should use an update script similar to the KDE gear one

 - https://www.plasma-mobile.org/info/plasma-mobile-gear-21-05/
 - https://github.com/NixOS/nixpkgs/pull/120514

samueldr committed:
aliases: Drop kalk (kalker)

The alias has been present for 40 days, the previous package was present
for not more than 3 days with the `kalk` name.

samueldr commented:
Previous force pushes:

 - Strictly rebases on top of nixos-unstable
 - Drops an alias we definitely can drop (#125555)

samueldr committed:
applications/plasma-mobile: Add fetch script

Used similarly to the KDE Gear scripts:

```
 $ ./maintainers/scripts/fetch-kde-qt.sh pkgs/applications/plasma-mobile/
```

samueldr committed:
applications/plasma-mobile: Init srcs.nix

Generated with the fetch script.

samueldr committed:
applications/plasma-mobile: Add empty plasma-mobile package set

samueldr committed:
qt5-packages.nix: Add plasmaMobileGear

Though it is empty, the package set is valid. Next changes introduce
derivations for packages we maintain.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29612,6 +29626,14 @@ in
 
   plasma-pass = libsForQt5.callPackage ../tools/security/plasma-pass { };
 
+  plasma-camera = libsForQt5.callPackage ../applications/graphics/plasma-camera/default.nix {
```
```suggestion
  plasma-camera = libsForQt5.callPackage ../applications/graphics/plasma-camera {
```

samueldr replied:
(Removed)

SuperSandro2000 commented on pkgs/applications/networking/plasma-dialer/default.nix:
```diff
@@ -0,0 +1,62 @@
+    libphonenumber
+    libpulseaudio
+    libqofono
+    protobuf # Needed by libphonenumber
```
Should this be propagated instead?

samueldr replied:
Good question, though note that all uses of `libphonenumber` add `protobuf` as an additional dependency.

I think either way it's a smell about the packaging of `libphonenumber`.

samueldr commented on pkgs/applications/plasma-mobile/fetch.sh:
```diff
@@ -0,0 +1 @@
+WGET_ARGS=( http://download.kde.org/stable/plasma-mobile/21.05 -A '*.tar.xz' )
```
I actually had to cheat to get the 21.05 release. Running the fetch script without changing the version downloads 21.05 *and* 21.06 packages here. Odd.

ttuegel replied:
That's strange... Here's what I know: `wget` is acting as a (not very smart) link scraper, looking for `<a>` tags with filenames that match the given pattern. This has broken pretty spectacularly in the past when, for example, KDE modified their website to insert most of the page content with JavaScript. I wonder if, at some point, the page source contained links for both versions? It only seems to have 21.05 now, though.

samueldr replied:
I made the same observations at the time. But I did not investigate more deeply. I guessed vaguely it might be following the link to the parent directory, but I don't get why it wouldn't download 21.05 with the 21.06.

samueldr commented:
Since Plasma Mobile is [now doing batched releases](https://www.plasma-mobile.org/info/plasma-mobile-gear-21-05/), this was refactored to work just like KDE Gear.

samueldr committed:
koko: init at 21.05

requires data that is not versioned:

 - https://invent.kde.org/graphics/koko#packaging

Snapshotted because... Obviously that won't do with Nix.

samueldr committed:
ktrip: init at 21.05

samueldr committed:
kpublictransport: init at 21.04.0

samueldr reviewed (commented):
[Comment GitHub forced me to leave to leave an in-line comment...]

siraben commented:
Result of `nixpkgs-review pr 125156` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages marked as broken and skipped:</summary>
  <ul>
    <li>libsForQt512.alligator</li>
    <li>libsForQt512.calindori</li>
    <li>libsForQt512.kalk</li>
    <li>libsForQt512.kclock</li>
    <li>libsForQt512.kirigami-addons</li>
    <li>libsForQt512.koko</li>
    <li>libsForQt512.kpublictransport</li>
    <li>libsForQt512.krecorder</li>
    <li>libsForQt512.ktrip</li>
    <li>libsForQt512.plasma-dialer</li>
    <li>libsForQt512.plasma-phonebook</li>
    <li>libsForQt512.spacebar</li>
    <li>libsForQt514.alligator</li>
    <li>libsForQt514.calindori</li>
    <li>libsForQt514.kalk</li>
    <li>libsForQt514.kclock</li>
    <li>libsForQt514.koko</li>
    <li>libsForQt514.kpublictransport</li>
    <li>libsForQt514.krecorder</li>
    <li>libsForQt514.ktrip</li>
    <li>libsForQt514.plasma-dialer</li>
    <li>libsForQt514.plasma-phonebook</li>
    <li>libsForQt514.spacebar</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>libsForQt514.kirigami-addons</li>
  </ul>
</details>
<details>
  <summary>15 packages built:</summary>
  <ul>
    <li>libsForQt5.alligator (libsForQt515.alligator ,plasma5Packages.alligator)</li>
    <li>libsForQt5.calindori (libsForQt515.calindori ,plasma5Packages.calindori)</li>
    <li>libsForQt5.kalk (libsForQt515.kalk ,plasma5Packages.kalk)</li>
    <li>libsForQt5.kclock (libsForQt515.kclock ,plasma5Packages.kclock)</li>
    <li>libsForQt5.kirigami-addons (libsForQt515.kirigami-addons ,plasma5Packages.kirigami-addons)</li>
    <li>libsForQt5.koko (libsForQt515.koko ,plasma5Packages.koko)</li>
    <li>libsForQt5.kpublictransport (libsForQt515.kpublictransport ,plasma5Packages.kpublictransport)</li>
    <li>libsForQt5.krecorder (libsForQt515.krecorder ,plasma5Packages.krecorder)</li>
    <li>libsForQt5.ktrip (libsForQt515.ktrip ,plasma5Packages.ktrip)</li>
    <li>libsForQt5.libqofono (libsForQt515.libqofono ,plasma5Packages.libqofono)</li>
    <li>libsForQt5.plasma-dialer (libsForQt515.plasma-dialer ,plasma5Packages.plasma-dialer)</li>
    <li>libsForQt5.plasma-phonebook (libsForQt515.plasma-phonebook ,plasma5Packages.plasma-phonebook)</li>
    <li>libsForQt5.spacebar (libsForQt515.spacebar ,plasma5Packages.spacebar)</li>
    <li>libsForQt512.libqofono</li>
    <li>libsForQt514.libqofono</li>
  </ul>
</details>

dotlambda commented:
@samueldr Any updates on this?

samueldr commented:
Reviews from @NixOS/qt-kde peeps.

I'm pretty sure nothing was left TODO, and that was why I asked for a review from @ttuegel. Mainly a review since it adds a new directory under `pkgs/applications` that uses `maintainers/scripts/fetch-kde-qt.sh`.

Mindavi commented:
I tried this out on the pinephone (tested kclock, krecorder and kalk).
![IMG_20211002_120413](https://user-images.githubusercontent.com/9799623/135711772-5605d5c3-4b38-4726-a8b3-b10b247df06e.jpg)
![Uploading IMG_20211002_120433.jpg…]()

Apart from krecorder missing some icons for the buttons (I had to click on them blindly), they do build on aarch64. Krecorder somehow doesn't show recordings that I save, but it's _probably_ not related to packaging.

Mindavi reviewed (approved):
Quickly skimmed over,  LGTM

samueldr commented:
@Mindavi just so it's clear, there's a list of caveats, and part of it has a *Icons wont work as expected* description. Icons missing outside of Plasma environments is a known issue, and squarely an ([ignored](https://bugs.kde.org/show_bug.cgi?id=435236)) upstream issue.

SuperSandro2000 commented on pkgs/applications/kde/kpublictransport.nix:
```diff
@@ -0,0 +1,13 @@
+{
+  mkDerivation, lib,
+  extra-cmake-modules,
+}:
+
+mkDerivation {
+  pname = "kpublictransport";
+  meta = {
+    license = [ lib.licenses.cc0 ];
+    maintainers = [ lib.maintainers.samueldr ];
+  };
+  nativeBuildInputs = [ extra-cmake-modules ];
+}
```
```suggestion
{ lib,
, mkDerivation
, extra-cmake-modules
}:

mkDerivation {
  pname = "kpublictransport";

  nativeBuildInputs = [ extra-cmake-modules ];

  meta = {
    license = [ lib.licenses.cc0 ];
    maintainers = [ lib.maintainers.samueldr ];
  };
}
```

SuperSandro2000 commented on pkgs/applications/plasma-mobile/default.nix:
```diff
@@ -0,0 +1,75 @@
+
+*/
+
+{
+  lib, libsForQt5, fetchurl,
+}:
```
```suggestion
{ lib
, libsForQt5
, fetchurl
}:
```
